### ISSUE TITLE
inadvertently using a non-distributed domain

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#5657]
### Summary of Problem

If the type is specified for a domain and it is initialized by an expression with a dmapped clause, it is surprising that the resulting domain is not distributed. The program below compiles and runs without error but does not give any indication that the line containing `dmapped` is problematic. In particular, that line is assigning a distributed domain to a non-distributed one. Perhaps a warning could be generated in this case.

### Steps to Reproduce

**Source Code:**

```chapel
  use BlockDist;
  config const n = 100;
  var myDomain: domain(2) = {1..n, 1..n} dmapped Block({1..n, 1..n});
  for d in myDomain {
    writeln(d.locale.id);
  }
```

**Comment by bradcray** [comment]
I don't agree that the line above is nonsensical, though I agree that it likely reflects a misunderstanding on the user's part, and one that real users have run afoul of at times.

Why I don't think it's nonsensical:  We permit domain assignment between domains of distinct distributions, and the definition is that the result is a copy of the index set, not of the implementation (its domain map / distribution).  So the statement above is legal and sensical even though it may not reflect the programmer's intention.  This property is why it's legal to do:

````chapel
var myDomain: domain(2) dmapped Block(...) = {1..n, 1..n};
````

even though the two domain maps don't match.

A language change we could consider would be to change the interpretation of:

````chapel
var myDomain: domain(2) = ...;
````

from:

````chapel
var myDomain: domain(2) dmapped DefaultDomainMap = ...
````

to:

````chapel
var myDomain: domain(2) dmapped [InheritDistFromInitExpr] = ...
````

which would be somewhat analogous to our treatment of formal array types like:

````chapel
proc foo(X: [1..n] real) { ... }
````

where we interpret this not as "the actual must be a default rectangular" but rather "the actual must have this index set and we'll inherit its domain map" (making the function generic w.r.t. domain map).

Alternatively, I feel like if we were to generate a warning in this context as you suggest ("you're initializing a non-distributed domain using a distributed domain -- do you mean to do that?  If not, add a dmapped clause to your type expression or remove the type expression altogether."), that would probably be sufficient as well.

**Comment by bradcray** [comment]
Thinking about this a bit more this morning (without realizing that you'd closed the issue), I increasingly think that the warning is the right way to go.  Reason being that I don't think the generic behavior we allow in formal argument lists (where the type qualifier is like a constraint) is something we'd want to have affect declarations (where the type specification says what the variable is).  For example, in a declaration like:

````chapel
var X: [1..n] real = myDistributedNElementArray;
````

I don't think we'd want to have 'X' become a distributed array just because the RHS is (even though we would permit an actual of the RHS pass to a formal of X's type).

**Comment by mppf** [comment]
@bradcray - I didn't close the issue - I created a closed PR that solves the issue. I closed the PR because it didn't seem to be a release priority but I think we could get it working any time we like, if that appeals to you in the near term.

In particular, whenever you have time to think about it, I'm interested in your feedback on the implementation approach in that PR. (Basically, it adds a flags "distributed dsi" and "non-distributed dsi" and marks BlockDom and DefaultRectangular etc with these flags as appropriate; then uses these flags to look for the problematic case in checkResolved()).

**Comment by bradcray** [comment]
Oh, you're right -- I mis-read the "closed" marker in the timeline above as being for this issue...

**Comment by benharsh** [comment]
Do we want this warning for initialization only, or for assignment as well?

**Comment by mppf** [comment]
I think only initialization - we're trying to find a way to warn on the code at the top that a user actually typed in. Assignment is more likely to be intentional.

**Comment by benharsh** [comment]
Resolved by #10791
